cli/ipc/ipc.go

```
@@ -0,0 +1,393 @@
+	// Ensure that message was consumed by the library
+	for i := 0; i < comRetryLimit; i++ {
+		curMsg, err := ipc.sender.size()
+		if err != nil {
+			return err
+		}
+		if curMsg != 0 {
+			time.Sleep(ipcComTimeout)
+		} else {
+			return nil
+		}
+	}
+	return ErrConsumerTimeout
+}
+
+// metadata in ipc frame
+func (ipc *ipcObj) frameMeta(remainBytes int) ([]byte, error) {
+	metadata, err := json.Marshal(metaRequest{Req: metaReqJsonPartial, Uniq: ipc.mqUniqId, Remain: remainBytes})
+	if err != nil {
+		return metadata, err
+	}
+	// Ensure that metadata is nul terminated
+	metadata = append(metadata, 0)
+	return metadata, nil
+}
+
+// prepareFrame preparet the scope message
+func (ipc *ipcObj) prepareFrame(remain int, scopeData []byte, offset int) ([]byte, int, error) {
+	frame, _ := ipc.frameMeta(remain)
+	maxPossibleDataLen := ipc.sender.cap - len(frame) - 1
+	if maxPossibleDataLen < 0 {
+		return frame, -1, errRequest
+	}
+
+	scopeFrameDataSize := maxPossibleDataLen
+	if remain < maxPossibleDataLen {
+		scopeFrameDataSize = remain
+	}
+
+	frameScopeData := scopeData[offset : offset+scopeFrameDataSize]
+
+	frame = append(frame, frameScopeData...)
+	frame = append(frame, 0)
+
+	return frame, scopeFrameDataSize, nil
+}
+
+// sendIpcRequest puts the request in message queue (the message can be splited by the frames)
+func (ipc *ipcObj) sendIpcRequest(scopeMsg []byte) error {
+	scopeMsgDataRemain := len(scopeMsg)
+	var scopeMsgOffset int
+	for scopeMsgDataRemain != 0 {
+
+		frame, dataSize, err := ipc.prepareFrame(scopeMsgDataRemain, scopeMsg, scopeMsgOffset)
+		if err != nil {
+			return err
+		}
+
+		scopeMsgOffset += dataSize
+		scopeMsgDataRemain -= dataSize
+
+		// Handling the last frame
+		//  - change the request type metaReqJsonPartial -> metaReqJson
+		if scopeMsgDataRemain == 0 {
+			frame = bytes.Replace(frame, []byte("req\":1"), []byte("req\":0"), 1)
+		}
+
+		// Retry mechanism ?
+		err = ipc.send(frame)
+		if err != nil {
+			return err
+		}
```

In response to the "Retry mechanism ?" comment...  I like your fail fast approach you have now.  Fail fast, and tear down the message queues.  Seems very straightforward, and like a good way to avoid all kinds of edge conditions, or at least making the error handling very consistent.  Don't know if you want to update this comment to reflect this. (?)

---

Good point I have removed the comment.